services/worker-manager/src/providers/aws.js

```diff
@@ -0,0 +1,306 @@
+      maxCapacity: config.launchConfig.MaxCount,
+      capacityPerInstance: config.capacityPerInstance,
+      running: workerPool.providerData[this.providerId].running,
+    });
+
+    const userData = {
+      rootUrl: this.rootUrl,
+      workerPoolId,
+      providerId: this.providerId,
+      workerGroup: this.providerId,
+    };
+
+    let spawned;
+    try {
+      spawned = await ec2.runInstances({
+        ...config.launchConfig,
+
+        // please make sure this string is no more than 1024 chars long
+        // https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-instance-metadata.html#instancedata-add-user-data
+        UserData: JSON.stringify(userData).toString('base64'),
+
+        MaxCount: config.launchConfig.MaxCount ? Math.max(toSpawn, config.launchConfig.MaxCount) : toSpawn,
+        MinCount: config.launchConfig.MinCount ? Math.min(toSpawn, config.launchConfig.MinCount) : toSpawn,
+        TagSpecifications: [
+          ...(config.launchConfig.TagSpecifications ? config.launchConfig.TagSpecifications : []),
+          {
+            ResourceType: 'instance',
+            Tags: [
+              {
+                Key: 'Provider',
+                Value: `wm-${this.providerId}`,
+              }, {
+                Key: 'Owner',
+                Value: workerPool.owner,
+              }],
+          },
+        ],
+      }).promise();
+    } catch (e) {
+      this.monitor.err(`🚨 Error calling AWS API: ${e}`);
+
+      return await workerPool.reportError({
+        kind: 'creation-error',
+        title: 'Instance Creation Error',
+        description: e.message,
+        notify: this.notify,
+        WorkerPoolError: this.WorkerPoolError,
+      });
+    }
+
+    return Promise.all(spawned.Instances.map(i => {
+      return this.Worker.create({
+        workerPoolId,
+        providerId: this.providerId,
+        workerGroup: this.providerId,
+        workerId: i.InstanceId,
+        created: new Date(),
+        expires: taskcluster.fromNow('1 week'),
+        state: this.Worker.states.REQUESTED,
+        providerData: {
+          region: config.region,
+          groups: spawned.Groups,
+          amiLaunchIndex: i.AmiLaunchIndex,
+          imageId: i.ImageId,
+          instanceType: i.InstanceType,
+          architecture: i.Architecture,
+          availabilityZone: i.Placement.AvailabilityZone,
+          privateIp: i.PrivateIpAddress,
+          owner: spawned.OwnerId,
+          state: i.State.Name,
+          stateReason: i.StateReason.Message,
+        },
+      });
+    }));
+  }
+
+  /**
+   * This method checks instance identity document authenticity
+   * If it's authentic it checks whether the data in it corresponds to the worker
+   *
+   * @param worker string
+   * @param workerPool string
+   * @param workerIdentityProof {document: string, signature: string}
+   * @returns {Promise<{expires: *}>}
+   */
+  async registerWorker({worker, workerPool, workerIdentityProof}) {
```

Comment:
From `docs/providers.md`: "Providers that wish to limit registration to once per worker should return an error message from this function if the worker is already RUNNING".  We definitely wish to do so with this provider!